rfcs/detecting_sensitive_textual_content/20230309-project_proposal_detecting_sensitive_textual_content.md

```diff
@@ -0,0 +1,379 @@
+- Could have different performance issues if decoding the hash is a heavy
+  operation for low-spec clients
+  - Looking at the code, I highly doubt this because it's a
+    [fast bit of maths](https://github.com/mad-gooze/fast-blurhash/blob/main/index.js)
+    that (due to the nature of BlurHash) ignores almost all the content of the
+    image rather than a Gaussian blur which needs to consider ever aspect of the
+    image in order to apply the blur
+
+It has the following upsides:
+
+- Does not require the client to download the image just to blur it
+  - If the client never unblurs an image, it will have saved all the bytes of
+    the image from being downloaded to the client
+- We can easily cache the hash in Redis to ameliorate performance over time
+  - If we do so, then we could use the hash for other things down the line (like
+    average colour filters) and add it directly to the documents via the same
+    ETL pipeline we'll develop for dead links, further ameliorating any API
+    performance issues
+- Does not rely on the client's ability to apply and render a Gaussian blur
+  transformation to potentially several images on a single page
+- Extremely minor and only barely relevant: If we calculated the BlurHash for
+  all images in the future in our re-crawler, we could use them as placeholders
+  rather than the skeleton boxes we currently use
+
+#### Recommendation
+
+If we can spare the time, I think BlurHash is the correct choice for the
+following reasons:
+
+1. While significantly more complex, it is not sufficiently complex to be
+   burdensome. The simplicity of the CSS filter makes BlurHash look monstrous,
+   but in the grand scheme of things, it's actually a relatively simple approach
+   to blurring the images.
+2. It avoids performance issues with relying on low-spec clients to apply
+   Gaussian blurs to potentially several images.
+3. BlurHash hashes have significant future value and all potential API level
+   performance issues can be ameliorated through caching hashes with Redis.
+
+The only significant downside that I think exists is the delay between
+unblurring and displayin7.29.1g the full image due to needing to download the
+full unblurred image. However, I think this is balanced out by the bytes saved
+over the network in _not_ sending images that might potentially never be
+unblurred.
+
+For reviewers: **What are you feelings on this issue and why?**. One alternative
+to going straight for BlurHash, if we wanted to get blurring out the door
+quickly is to treat BlurHash as a future "nice to have" and use CSS blurring in
+the meantime. This would allow us to get a fully working feature slightly more
+quickly. I'm not totally convinced it'd be _that_ much faster though. Like I
+said above, BlurHash looks complicated, but really only in comparison to how
+simple the CSS approach is. Anything would look complicated when compared to
+just adding a CSS filter style to something.
```

Comment:
What if we had a single generic blurred image instead of trying to blur the one that needs it? This way, the user would not be able to discern what's behind the blur, but is it really necessary?

Reply:
> This way, the user would not be able to discern what's behind the blur

I think it's desirable to be able to sort of "guess" at what's behind the blur. For example, if you notice a lot of red, maybe the image includes a large amount of blood or something you'd rather avoid seeing, even though some sensitive results might be of interest for your query.

On the other hand, there are issues with this assumption as well, as in the case of this Mastodon issue: https://github.com/mastodon/mastodon/issues/10729

Our context might be different from Mastodon, though we do include Flickr, which is social media, and may have the same issues the Mastodon users are describing in that issue.

If we used a single generic blurred image, how would we choose it? We decided against using anything associated with Openverse (like the logo) to avoid implicitly associating Openverse with sensitive content. Are there other reasonable options that would potentially mislead people? Maybe a blank box or some other solid colour would be appropriate :thinking:

Reply:
Oh, I'll also note that the issue is slightly different from Mastodon in that on Mastodon images are often marked with author-composed "content warnings". We won't have those, at least not for the time being (until we implement some hypothetical categorisation of sensitive terms, for example. In our case, the blurred image would be an even more valuable signal for someone deciding whether a result might be relevant to their query.

Reply:
FWIW Pixabay places a black overlay with text "Adult Content" (more or less like a Mastodon content warning) that toggles off when a user selects option to remove safe search filters. No, it does not provide a hint to the image.

Reply:
That's an interesting alternative, thanks cogdog. We could do something similar with "Sensitive Content" for the text if folks felt it was a better approach.